File: frontend/types/product.ts
export type ProductCategory = 'FRUIT' | 'VEGETABLE' | 'GRAIN' | 'OTHER';
export type ProductStatus = 'NORMAL' | 'LOW_STOCK' | 'OUT_OF_STOCK' | 'SCHEDULED';
export type ProductUnit = 'kg' | 'box' | 'piece' | 'bag';

export interface Product {
  id: string;
  seller_id: string;
  name: string;
  category: ProductCategory;
  origin: string | null;
  spec: string | null;
  unit: ProductUnit;
  price_per_unit: number;
  stock_quantity: number;
  min_order_qty: number;
  status: ProductStatus;
  description: string | null;
  image_url: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProductCreate {
  name: string;
  category: ProductCategory;
  origin?: string;
  spec?: string;
  unit: ProductUnit;
  price_per_unit: number;
  stock_quantity?: number;
  min_order_qty?: number;
  description?: string;
}

export interface ProductUpdate {
  name?: string;
  category?: ProductCategory;
  origin?: string;
  spec?: string;
  unit?: ProductUnit;
  price_per_unit?: number;
  stock_quantity?: number;
  min_order_qty?: number;
  status?: ProductStatus;
  description?: string;
  image_url?: string;
}

// ===========================================
// 상품 상세 페이지 보강 응답 (B.1, 2026-05-04)
// 백엔드 schemas/product.py 의 ProductDetailResponse / ProductMinimal 와 동기화.
// ===========================================
/**
 * 같은 판매자의 다른 상품 — 상세 페이지 측면 노출용 미니 카드.
 * Product 의 일부 필드만 사용 (description / origin / spec 제외).
 */
export interface ProductMinimal {
  id: string;
  name: string;
  category: ProductCategory;
  unit: ProductUnit;
  price_per_unit: number;
  stock_quantity: number;
  status: ProductStatus;
  image_url: string | null;
}

/**
 * 거래처 관계 상태 (BUYER 가 본인 ↔ 판매자 partners.status 를 본 값).
 * - 'ACTIVE'           : 거래 중
 * - 'PENDING_OUTGOING' : 본인이 신청 후 대기
 * - 'PENDING_INCOMING' : 상대가 신청 → 본인 응답 대기
 * - 'INACTIVE'         : 비활성/거절 이력
 * - null               : 거래처 등록 이력 없음 (신규)
 */
export type PartnerRelationshipStatus =
  | 'ACTIVE'
  | 'PENDING_OUTGOING'
  | 'PENDING_INCOMING'
  | 'INACTIVE';

/**
 * GET /products/{id}/detail 응답.
 * 기존 Product 필드 + 판매자 join + (BUYER 한정) 거래 통계 + 같은 판매자 다른 상품.
 */
export interface ProductDetailResponse extends Product {
  seller_name: string | null;
  seller_company: string | null;
  partner_relationship_status: PartnerRelationshipStatus | null;
  previous_order_count: number;
  completed_order_count: number;
  other_seller_products: ProductMinimal[];
}

File: backend/app/schemas/product.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class ProductCreate(BaseModel):
    name: str
    category: str
    origin: Optional[str] = None
    spec: Optional[str] = None
    unit: str
    price_per_unit: int
    stock_quantity: int = 0
    min_order_qty: int = 1
    description: Optional[str] = None


class ProductUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    origin: Optional[str] = None
    spec: Optional[str] = None
    unit: Optional[str] = None
    price_per_unit: Optional[int] = None
    stock_quantity: Optional[int] = None
    min_order_qty: Optional[int] = None
    status: Optional[str] = None
    description: Optional[str] = None
    image_url: Optional[str] = None


class ProductResponse(BaseModel):
    id: UUID
    seller_id: UUID
    name: str
    category: str
    origin: Optional[str] = None
    spec: Optional[str] = None
    unit: str
    price_per_unit: int
    stock_quantity: int
    min_order_qty: Optional[int] = None
    status: str
    description: Optional[str] = None
    image_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ===========================================
# 상품 상세 페이지용 보강 응답 (B.1, 2026-05-04)
# ===========================================
class ProductMinimal(BaseModel):
    """같은 판매자의 다른 상품 미니 카드용 — 상세 페이지 사이드 노출."""

    id: UUID
    name: str
    category: str
    unit: str
    price_per_unit: int
    stock_quantity: int
    status: str
    image_url: Optional[str] = None

    model_config = {"from_attributes": True}


class ProductDetailResponse(ProductResponse):
    """GET /products/{id}/detail 전용 응답.

    - seller_name / seller_company: users 테이블 조인.
    - partner_relationship_status: 현재 사용자(BUYER)와 판매자 사이 partners 관계 상태.
        값: 'ACTIVE' | 'PENDING_OUTGOING' | 'PENDING_INCOMING' | 'INACTIVE' | None
        (None 이면 거래처 등록 이력 자체 없음)
    - previous_order_count: 본인과 이 판매자 사이의 주문 총 건수.
        soft-deleted/CANCELLED 제외. completed_order_count 는 그 중 COMPLETED 만.
        SELLER 가 본인 상세를 조회하면 두 카운트 모두 0 (의미 없음).
    - other_seller_products: 같은 판매자의 다른 상품 (본 상품 제외, 최대 4개,
        OUT_OF_STOCK 우선순위 낮춤, status != OUT_OF_STOCK 우선).
    """

    seller_name: Optional[str] = None
    seller_company: Optional[str] = None
    partner_relationship_status: Optional[str] = None
    previous_order_count: int = 0
    completed_order_count: int = 0
    other_seller_products: list[ProductMinimal] = []

    model_config = {"from_attributes": True}
